src/disco/shred/fd_fec_resolver.c

```diff
@@ -0,0 +1,527 @@
+  ulong complete_depth = resolver->complete_depth;
+  ulong done_depth     = resolver->done_depth;
+
+  int lg_depth = fd_ulong_find_msb( depth );
+  ulong done_map_cnt = fd_tcache_map_cnt_default( done_depth );
+
+  FD_SCRATCH_ALLOC_INIT( l, shmem );
+  /*     self       */ FD_SCRATCH_ALLOC_APPEND( l, FD_FEC_RESOLVER_ALIGN,  sizeof(fd_fec_resolver_t)                       );
+  void * done        = FD_SCRATCH_ALLOC_APPEND( l, FD_TCACHE_ALIGN,        fd_tcache_footprint( done_depth, done_map_cnt ) );
+  void * curr        = FD_SCRATCH_ALLOC_APPEND( l, ctx_map_align(),        ctx_map_footprint( lg_depth+1 )                 );
+  void * free        = FD_SCRATCH_ALLOC_APPEND( l, freelist_align(),       freelist_footprint( depth+partial_depth+1UL )   );
+  void * cmplst      = FD_SCRATCH_ALLOC_APPEND( l, freelist_align(),       freelist_footprint( complete_depth+1UL  )       );
+  void * bmfree      = FD_SCRATCH_ALLOC_APPEND( l, bmtrlist_align(),       bmtrlist_footprint( depth+1UL )                 );
+  FD_SCRATCH_ALLOC_FINI( l, FD_FEC_RESOLVER_ALIGN );
+
+  resolver->done             = fd_tcache_join( done   );
+  resolver->curr_map         = ctx_map_join  ( curr   );
+  resolver->free_list        = freelist_join ( free   );
+  resolver->complete_list    = freelist_join ( cmplst );
+  resolver->bmtree_free_list = bmtrlist_join ( bmfree );
+  fd_sha512_join( resolver->sha512 );
+
+  return resolver;
+}
+
+int fd_fec_resolver_add_shred( fd_fec_resolver_t    * resolver,
+                               fd_shred_t   const   * shred,
+                               ulong                  shred_sz,
+                               uchar  const   * leader_pubkey,
+                               fd_fec_set_t const * * out_fec_set,
+                               fd_shred_t   const * * out_shred ) {
+  /* Unpack variables */
+  fd_tcache_t  * done          = resolver->done;
+  ulong          done_depth    = fd_tcache_depth       ( done );
+  ulong          done_map_cnt  = fd_tcache_map_cnt     ( done );
+  ulong        * done_oldest   = fd_tcache_oldest_laddr( done );
+  ulong        * done_ring     = fd_tcache_ring_laddr  ( done );
+  ulong        * done_map      = fd_tcache_map_laddr   ( done );
+
+  ulong partial_depth = resolver->partial_depth;
+
+  fd_fec_set_t * * free_list        = resolver->free_list;
+  fd_fec_set_t * * complete_list    = resolver->complete_list;
+  void         * * bmtree_free_list = resolver->bmtree_free_list;
+  set_ctx_t    *   curr_map         = resolver->curr_map;
+
+  fd_reedsol_t * reedsol       = resolver->reedsol;
+  fd_sha512_t  * sha512        = resolver->sha512;
+
+  /* Invariants:
+      * no key is in both the done tcache and the current map
+      * each set pointer provided to the new function is in exactly one
+          of curr_map, freelist, or complete_list
+      * bmtree_free_list has exactly partial_depth fewer elements than
+          freelist
+   */
+
+  int dup = 0;
+  (void)dup;
```

Comment:
Why (void) ?

Reply:
It was to prevent a variable read but never written warning. It's gone now with the switch from tcache to map.